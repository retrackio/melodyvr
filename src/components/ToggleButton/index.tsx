import React, { useState, useMemo } from 'react';
import * as Styled from './style';

interface Props<V> {
  id: string;
  leftValue: V;
  leftLabel: string | React.ReactNode;
  rightValue: V;
  rightLabel: string | React.ReactNode;
  value: V;
  onChange: (value: V) => void;
}

function ToggleButton<V = string>({
  value,
  leftLabel,
  leftValue,
  rightLabel,
  rightValue,
  onChange,
  id,
}: Props<V>) {
  const [checked, setChecked] = useState<boolean>(value === rightValue);
  const handleChange = useMemo(
    () => ({ target }: React.ChangeEvent<HTMLInputElement>) => {
      setChecked(target.checked);
      onChange(target.checked ? rightValue : leftValue);
    },
    [],
  );
  return (
    <Styled.Container>
      <Styled.Label>{leftLabel}</Styled.Label>
      <Styled.CheckBoxWrapper>
        <Styled.CheckBox
          id={id}
          type="checkbox"
          checked={checked}
          onChange={handleChange}
        />
        <Styled.CheckBoxLabel htmlFor={id} />
      </Styled.CheckBoxWrapper>
      <Styled.Label>{rightLabel}</Styled.Label>
    </Styled.Container>
  );
}

export default ToggleButton;
