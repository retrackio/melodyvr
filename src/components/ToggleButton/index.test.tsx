import React from 'react';
import { shallow } from 'enzyme';
import ToggleButton from './';

describe('<ToggleButton />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <ToggleButton<string>
        id="test1"
        value="a"
        leftLabel="A"
        leftValue="a"
        rightLabel="B"
        rightValue="b"
        onChange={jest.fn()}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('calls the onChange callback with the correct value', () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <ToggleButton<string>
        id="test1"
        value="a"
        leftLabel="A"
        leftValue="a"
        rightLabel="B"
        rightValue="b"
        onChange={onChange}
      />,
    );
    const checkbox = wrapper.find('[type="checkbox"]');
    checkbox.simulate('change', { target: { checked: true } });
    expect(onChange).toHaveBeenCalledWith('b');
    checkbox.simulate('change', { target: { checked: false } });
    expect(onChange).toHaveBeenCalledWith('a');
  });
});
