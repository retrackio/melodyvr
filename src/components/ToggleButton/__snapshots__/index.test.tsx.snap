// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleButton /> renders correctly 1`] = `
<style__Container>
  <style__Label>
    A
  </style__Label>
  <style__CheckBoxWrapper>
    <style__CheckBox
      checked={false}
      id="test1"
      onChange={[Function]}
      type="checkbox"
    />
    <style__CheckBoxLabel
      htmlFor="test1"
    />
  </style__CheckBoxWrapper>
  <style__Label>
    B
  </style__Label>
</style__Container>
`;
